### PR TITLE
Rewrite collectCommonConjunctions without recursion

### PR DESCRIPTION
**Pull request by mariadb-AndreyPiskunov**

### CODE REVIEW

File: dbcon/execplan/rewrites.cpp [hunk 1]
```diff
@@ -83,6 +80,13 @@ enum class ChildType
   Leave
 };
 
+enum class GoTo
+{
```
Comment (drrtuy) [inline dbcon/execplan/rewrites.cpp]: Is it possible to re-use GoTo enum defined in parstree.h?

File: dbcon/execplan/rewrites.cpp [hunk 2]
```diff
@@ -110,44 +114,104 @@ bool simpleFiltersCmp(const SimpleFilter* left, const SimpleFilter* right)
...
+void advanceSetUp(std::vector<StackFrameWithSet>& stack, CommonContainer& accumulator)
+{
+  if (stack.size() == 1)
+    accumulator = stack.back().localset;
+  else
   {
-    // we want to collect it if it is a child of and node and or node was met before
-    if (castToFilter(root))
+    auto sz = stack.size();
+    if (operatorType(stack.at(sz - 2).node) == OP_OR)
```
Comment (drrtuy) [inline dbcon/execplan/rewrites.cpp:132]: Do we check the invariant that there is something at(sz-2)? If not std::optional looks like a natural fit here.

**Review by drrtuy — COMMENTED** [review]
LGTM. There are couple questions in the patch.

File: dbcon/execplan/rewrites.cpp (hunk 1 above)
Comment (mariadb-AndreyPiskunov) [reply dbcon/execplan/rewrites.cpp]: Yeah, this was created after, but maybe now it makes sense to reuse.

File: dbcon/execplan/rewrites.cpp (hunk 2 above)
Comment (mariadb-AndreyPiskunov) [reply dbcon/execplan/rewrites.cpp:132]: stack size is always > 0  in the while loop outside. and the = 1 case is checked on line 132.

File: dbcon/execplan/rewrites.cpp (hunk 1 above)
Comment (mariadb-AndreyPiskunov) [reply dbcon/execplan/rewrites.cpp]: Done.

**Review by mariadb-LeonidFedorov — APPROVED** [review]

**Review by drrtuy — APPROVED** [review]

**Review by mariadb-LeonidFedorov — CHANGES_REQUESTED** [review]
You should check the new test for MTR proposed, it's not passing on CI

File: dbcon/execplan/objectreader.cpp [hunk 3]
```diff
@@ -195,33 +195,102 @@ void ObjectReader::writeParseTree(const ParseTree* tree, messageqcpp::ByteStream
...
 
   ret = new ParseTree();
-  ret->left(createParseTree(b));
-  ret->right(createParseTree(b));
-  ret->data(createTreeNode(b));
+  stack.emplace_back(ret);
+  while (!stack.empty())
+  {
+    auto [node, dir] = stack.back();
+    if (dir == ParseTree::GoTo::Left)
```
Comment (drrtuy) [inline dbcon/execplan/objectreader.cpp:248]: There is a pattern matching lib in the repo. Use it next time if you decide to create a enum to follow FP patterns.

File: dbcon/execplan/parsetree.h [hunk 4]
```diff
@@ -378,65 +394,174 @@ inline ParseTree::ParseTree(const ParseTree& rhs)
...
+        if (node->fRight != nullptr)
+        {
+          stack.emplace_back(node->fRight);
+        }
+      }
+      else
+      {
+        if (stack.size() == 1)
+        {
+          node->fLeft = nullptr;
```
Comment (drrtuy) [inline dbcon/execplan/parsetree.h:433]: else scope shares these two lines so you can put them before the if. Both if and else shares `stack,pop_back()` so you can put them outside if IMHO.

**Review by drrtuy — APPROVED** [review]
LGTM except minor boilerplate part that could be fixed before the merge.

**Review by mariadb-LeonidFedorov — APPROVED** [review]

**Pull request merged**